### PR TITLE
SSL certificate verification

### PR DESCRIPTION
**Pull request by daviesian**
This PR is an improved and extended version of #8252, which adds support for the `cert_reqs` and `ca_certs` arguments to `ssl.wrap_socket`. PR #8252 only verifies the server certificate when doing mutual TLS, but this version verifies the server certificate whenever one is provided and `cert_reqs` is set appropriately. This is implemented in the first commit - ~6dd27eef03f223c9e430254262a8d3f4ab8402a1~ 91d30daed3b8ffcf035f1b0cd7d5e334af195d41, relying heavily on #8252 (thanks @Carglglz!).

This PR also verifies certificate time validity on the RP2 port, making use of the RTC. If `cert_reqs` is set to `ssl.CERT_REQUIRED`, certificates will only be verified successfully if they are valid at the current time. This is implemented in the second commit - ~e306e2afded979957873db324b72243f2b38cfa6~ 16649390ac32dd0fac03edab16b626aa2de2cf5d.

Of course, this requires the RTC to be set correctly, so this PR also includes an implementation of the `ntptime` module, copied and modified from the ESP8266 port. This is implemented in the third commit - ~b97a3423aa6d828fb191e6f5c43daabbcbc9655b~ 4b4843f45eed6d63a70b83a009e7c049c7602361. Before making SSL connections, developers should call `ntptime.settime()` or set the RTC to the correct time by some other means.

I'm not expecting this PR to be directly merged as-is, but would welcome feedback on the wisdom (or otherwise) of the various choices I've proposed. Namely:
- Moving `ca_certs` handling logic out one level, so it happens whenever the `ca_certs` argument is provided. @Carglglz - does this sound sensible?
- Unconditionally adding `MBEDTLS_BASE64_C` and `MBEDTLS_PEM_PARSE_C` to the RP2 port, making it easy for developers to provide `ca_certs`.
- Always verifying the valid/expired status of server certificates based on the RTC. With this implementation, there is no way to verify everything _apart_ from the time on a certificate, which feels correct to me. A certificate is either valid or it isn't.

If this is broadly acceptable then I'd be happy to polish this PR into something suitable for merging.

### CODE REVIEW

**Comment by iabdalkader** [comment]
This is great, thanks for this! One comment though, I think `ntptime.py` should be removed from `esp` port and put in a common place (ex `/extmod), this way it's easy to reuse this from `stm32` and other ports.

**Comment by daviesian** [comment]
Great idea, I'll do that and update the PR!

**Comment by daviesian** [comment]
Actually, it's not quite that simple, because the different boards each use a different epoch (the RP2 uses 1970-1-1, the ESP8266 uses 2000-1-1). I could easily create a per-board module called something like `system_epoch` so `ntptime` could do `from system_epoch import EPOCH`, but that feels messy. @iabdalkader can you think of a cleaner solution?

**Comment by iabdalkader** [comment]
I'm not sure, maybe just define the epoch in time modules ? For example in 
```diff
diff --git a/ports/rp2/modutime.c b/ports/rp2/modutime.c
index ac291c893..312381b14 100644
--- a/ports/rp2/modutime.c
+++ b/ports/rp2/modutime.c
@@ -117,6 +117,11 @@ STATIC const mp_rom_map_elem_t mp_module_time_globals_table[] = {
     { MP_ROM_QSTR(MP_QSTR_ticks_cpu), MP_ROM_PTR(&mp_utime_ticks_cpu_obj) },
     { MP_ROM_QSTR(MP_QSTR_ticks_add), MP_ROM_PTR(&mp_utime_ticks_add_obj) },
     { MP_ROM_QSTR(MP_QSTR_ticks_diff), MP_ROM_PTR(&mp_utime_ticks_diff_obj) },
+    #if MICROPY_EPOCH_IS_1970
+    { MP_ROM_QSTR(MP_QSTR_epoch), MP_ROM_INT(1970) },
+    #else
+    { MP_ROM_QSTR(MP_QSTR_epoch), MP_ROM_INT(2000) },
+    #endif
 };
```

Alternatively, could also just add an arg ntptime.time ?
```diff
diff --git a/ports/esp8266/modules/ntptime.py b/ports/esp8266/modules/ntptime.py
index dd07e46f1..cddc96a4d 100644
--- a/ports/esp8266/modules/ntptime.py
+++ b/ports/esp8266/modules/ntptime.py
@@ -14,7 +14,7 @@ NTP_DELTA = 3155673600
 host = "pool.ntp.org"

-def time():
+def time(epoch = 1970):
     NTP_QUERY = bytearray(48)
     NTP_QUERY[0] = 0x1B
     addr = socket.getaddrinfo(host, 123)[0][-1]
@@ -26,7 +26,7 @@ def time():
     finally:
         s.close()
     val = struct.unpack("!I", msg[40:44])[0]
-    return val - NTP_DELTA
+    return val - (NTP_DELTA if epoch = 1970 else delta_2000)
```

**Comment by Carglglz** [comment]
@daviesian Nice to see that this gets some attention 👍🏼 

> PR https://github.com/micropython/micropython/pull/8252 only verifies the server certificate when doing mutual TLS, but this version verifies the server certificate whenever one is provided and cert_reqs is set appropriately. This is implemented in the first commit - https://github.com/micropython/micropython/commit/6dd27eef03f223c9e430254262a8d3f4ab8402a1,

> Moving ca_certs handling logic out one level, so it happens whenever the ca_certs argument is provided. @Carglglz - does this sound sensible?

Yes probably is better to move it out one level, it makes sense, I don't know why I didn't that in the first place...

RTC check for expiration date is a nice addition as well as PEM format compatibility. 
Adding this to other ports would be nice too. 👍🏼 

IMO I would convert this to a draft and then split this in 2 or 3 PRs which may be easier to review/discuss/merge , one PR could be just rp2 port and the other esp/stm32/rest... 🤔

**Comment by daviesian** [comment]
Thanks @iabdalkader and @Carglglz for your comments and suggestions. I have now updated the branch to address the issue with the duplicated `ntptime` module in the cleanest way I could manage. You'll see that `ntptime` is now sourced from the `extmod/ntptime_py` directory, with an optional `ntptime.epoch_override` module that should be added in ports that have `MICROPY_EPOCH_IS_1970` defined. You can see an example of that in [`ports/rp2/netmodules/ntptime/epoch_override.py`](https://github.com/micropython/micropython/pull/8854/commits/b952c18d35f618276d1f80c4dbf87fc31c42662f#diff-cbe6e516239fb3112de2bc0a3ce54b8cc66ef2a818593d79e773c5b2e21b3730).

I tried a bunch of different options for overriding the epoch in some ports, and this seemed the cleanest. I have added a README file in the common `ntptime` package in `extmod`, so future developers should find it fairly easily.

**Recap**

This PR is an improved version of #8252, which adds full TLS certificate verification to the RP2 port, and factors out the `ntptime` module for use in multiple ports. The checks are now passing, so I think this is good to go. Once we've agreed the right approach, I'll be happy to create additional PRs to enable this TLS validation functionality on the other ports, as suggested by @Carglglz.

@dpgeorge, are you happy with this?

**Comment by dpgeorge** [comment]
Yes this looks really good!

(See other previous attempt #4419.)

File: extmod/modussl_mbedtls.c [hunk 1]
```diff
@@ -173,7 +175,7 @@ STATIC mp_obj_ssl_socket_t *socket_new(mp_obj_t sock, struct ssl_args *args) {
     mbedtls_ctr_drbg_init(&o->ctr_drbg);
     #ifdef MBEDTLS_DEBUG_C
     // Debug level (0-4) 1=warning, 2=info, 3=debug, 4=verbose
-    mbedtls_debug_set_threshold(0);
+    mbedtls_debug_set_threshold(3);
```
Comment (dpgeorge) [inline extmod/modussl_mbedtls.c:178]: why this change?

File: extmod/modussl_mbedtls.c [hunk 2]
```diff
@@ -191,7 +193,11 @@ STATIC mp_obj_ssl_socket_t *socket_new(mp_obj_t sock, struct ssl_args *args) {
         goto cleanup;
     }
 
-    mbedtls_ssl_conf_authmode(&o->conf, MBEDTLS_SSL_VERIFY_NONE);
+
+    int certreq = (args->cert_reqs.u_int);
+    mbedtls_ssl_conf_authmode(&o->conf, certreq);
```
Comment (dpgeorge) [inline extmod/modussl_mbedtls.c:198]: since `certreq` is not reused, cleaner to write `mbedtls_ssl_conf_authmode(&o->conf, args->cert_reqs.u_int);`

File: extmod/ntptime_py/ntptime/__init__.py [hunk 3]
```diff
@@ -6,9 +6,11 @@
...
 except:
     import struct
-
-# (date(2000, 1, 1) - date(1900, 1, 1)).days * 24*60*60
-NTP_DELTA = 3155673600
+try:
+    # The epoch_override module will only be present on builds where the epoch year is not 2000
+    from ntptime.epoch_override import EPOCH_YEAR
+except:
+    EPOCH_YEAR = 2000
```
Comment (dpgeorge) [inline extmod/ntptime_py/ntptime/__init__.py:13]: It's possible to detect the Epoch used by a system at runtime using `time.gmtime(0)[0]`.  That's the year of the Epoch (1970 or 2000).  I think it'll be much simpler to use that mechanism to detect it.

File: ports/esp8266/boards/GENERIC_512K/manifest.py [hunk 4]
```diff
@@ -1,5 +1,4 @@
 freeze("$(BOARD_DIR)", "_boot.py", opt=3)
-freeze("$(PORT_DIR)/modules", ("apa102.py", "ntptime.py", "port_diag.py"))
```
Comment (dpgeorge) [inline ports/esp8266/boards/GENERIC_512K/manifest.py:2]: this looks like it removes things that don't have anything to do with this PR?

File: extmod/modussl_mbedtls.c (hunk 1 above)
Comment (daviesian) [reply extmod/modussl_mbedtls.c:178]: Debugging is already normally disabled by the use of `MBEDTLS_DEBUG_C`, so there's no need to also set the default level to `0`. If someone is enabling debugging, they probably want _some_ amount of useful output by default. This just seemed like a sensible default level.

File: extmod/modussl_mbedtls.c (hunk 2 above)
Comment (daviesian) [reply extmod/modussl_mbedtls.c]: Good catch - [updated](https://github.com/daviesian/micropython/blob/ssl-certificate-verification/extmod/modussl_mbedtls.c#L196)!

File: extmod/ntptime_py/ntptime/__init__.py (hunk 3 above)
Comment (daviesian) [reply extmod/ntptime_py/ntptime/__init__.py]: Why didn't I think of that?! 🙈
Thank you, this makes it [**so** much simpler](https://github.com/micropython/micropython/blob/4b4843f45eed6d63a70b83a009e7c049c7602361/extmod/ntptime.py).

File: ports/esp8266/boards/GENERIC_512K/manifest.py (hunk 4 above)
Comment (daviesian) [reply ports/esp8266/boards/GENERIC_512K/manifest.py:2]: I needed to remove `ntptime.py` from this list, then spotted that the manifest in the parent directory was already freezing everything in `$(PORT_DIR)/modules`, so this entire line was already redundant. At that point, it seemed a bit silly to just remove `ntptime.py` rather than the whole line.

**Comment by daviesian** [comment]
@dpgeorge Thanks very much for your very sensible suggestions, I've pushed an updated implementation. Is there anything further you'd like me to address?

File: extmod/ntptime.py [hunk 5]
```diff
@@ -26,14 +25,26 @@ def time():
...
+    EPOCH_YEAR = time.gmtime(0)[0]
+    if EPOCH_YEAR == 2000:
+        # (date(2000, 1, 1) - date(1900, 1, 1)).days * 24*60*60
+        NTP_DELTA = 3155673600
+    elif EPOCH_YEAR == 1970:
+        # (date(1970, 1, 1) - date(1900, 1, 1)).days * 24*60*60
+        NTP_DELTA = 2208988800
+    else:
+        raise Exception(
+            f"Unsupported EPOCH_YEAR provided in this port of MicroPython: {EPOCH_YEAR}. Cannot interpret NTP time."
```
Comment (dpgeorge) [inline extmod/ntptime.py:38]: I think it's better not to use f-strings, they are a relatively new feature to uPy and not enabled on all ports/boards.  Also this error message could be made a bit shorter, eg `"unsupported Epoch {}".format(EPOCH_YEAR)`

File: ports/esp8266/boards/GENERIC_512K/manifest.py (hunk 4 above)
Comment (dpgeorge) [reply ports/esp8266/boards/GENERIC_512K/manifest.py:2]: OK, sounds good.

File: extmod/modussl_mbedtls.c (hunk 1 above)
Comment (dpgeorge) [reply extmod/modussl_mbedtls.c:178]: OK

File: ports/rp2/mbedtls/mbedtls_config.h [hunk 6]
```diff
@@ -94,6 +94,8 @@
 #define MBEDTLS_X509_CRT_PARSE_C
 #define MBEDTLS_X509_USE_C
 
+#define MBEDTLS_BASE64_C
+#define MBEDTLS_PEM_PARSE_C
```
Comment (dpgeorge) [inline ports/rp2/mbedtls/mbedtls_config.h:98]: please split this change into the rp2 commit (so that there's on commit that just touches `extmod/modussl_mbedtls.c`)

**Comment by dpgeorge** [comment]
I'm inclined to merge #8252 which does implement the core `ca_certs` and `cert_reqs` arguments to `wrap_socket()`.  Because that PR was first and to give attribution to @Carglglz .  There's also #5998 but #8252 is an improved version of #5998.

This PR could then contribute 1) ntptime generalisation; 2) rp2 improvements.

**Comment by daviesian** [comment]
That sounds like an excellent plan - I would have based this PR off that one if PRs-on-PRs didn't make my head explode. I'll watch for that one being merged, then rebase this one. Thanks!

**Comment by dpgeorge** [comment]
@daviesian can you please rebase on master now that `cert_reqs` and `cadata` are implemented?

File: extmod/ntptime.py (hunk 5 above)
Comment (daviesian) [reply extmod/ntptime.py]: No problem - done!

File: ports/rp2/mbedtls/mbedtls_config.h (hunk 6 above)
Comment (daviesian) [reply ports/rp2/mbedtls/mbedtls_config.h:97]: Done!

**Comment by daviesian** [comment]
@dpgeorge Sure thing, this is now rebased on `master`. Just shout if you'd like any further tweaks, otherwise I think this is ready to merge.

File: ports/rp2/mbedtls/mbedtls_config.h (hunk 6 above)
Comment (dpgeorge) [reply ports/rp2/mbedtls/mbedtls_config.h:97]: Thanks.

But another point of discussion is whether enabling PEM parsing is a good idea.  If it's enabled here then it sets a precedence that other ports should follow, otherwise writing portable code is hard (eg using a PEM in code on rp2 won't then work on an stm32 board).

PEM parsing adds a lot of unnecessary overhead (code size and RAM and time), it's much simpler to use DER format.  PEM is essentially base-64 encoded DER with header/footer data.  So to parse PEM means allocating a chunk of memory to hold the DER and then base-64 decoding the PEM into that buffer, and then using the DER (eg parsing it).

I think for MicroPython we should standardise on using (and documenting how to use) DER format only.  I'd be interested to hear your thoughts on this.

In order not to hold up this PR any longer, would be good not enable PEM here, and instead discuss it more widely.

**Comment by iabdalkader** [comment]
Can these changes be propagated to other ports, especially the `stm32` ? It will save more PRs and reviews later.

**Comment by daviesian** [comment]
That would be great, but I don't have knowledge of, or access to, any other hardware. Once this is merged, it should be as simple as implementing the time hook in [this commit](https://github.com/micropython/micropython/pull/8854/commits/d187bb032b43fb5d85dab92d5a827332f9dafaa6) for other ports, so the PR/review process should be pretty minimal for those.

File: ports/rp2/mbedtls/mbedtls_config.h (hunk 6 above)
Comment (dpgeorge) [reply ports/rp2/mbedtls/mbedtls_config.h:97]: @daviesian any thoughts on this point?  My suggestion is to not add these 2 lines in this PR.

**Comment by daviesian** [comment]
@dpgeorge Is there anything else needed here before merging? It would be great to get this into master!

File: ports/rp2/mbedtls/mbedtls_config.h (hunk 6 above)
Comment (dpgeorge) [reply ports/rp2/mbedtls/mbedtls_config.h:97]: @daviesian ping

**Comment by dpgeorge** [comment]
> Is there anything else needed here before merging?

See inline comment above.

File: ports/rp2/mbedtls/mbedtls_config.h (hunk 6 above)
Comment (daviesian) [reply ports/rp2/mbedtls/mbedtls_config.h]: Sorry for missing this! You're quite right that this doesn't really belong here, although I do think I would be in favour of including it by default. Removed!

**Comment by daviesian** [comment]
Sorry for missing the inline comment - I clearly need to improve my Github Notification settings! Anyway, I've now removed those unnecessary `#define`s.

**Comment by dpgeorge** [comment]
> Sorry for missing the inline comment

Not a problem.  GitHub has a lot of notifications, and it's easy to miss them.

But thanks for updating.  Rebased and merged in b560b9fe715e293caef80533d61d8c3b179b0339 through 1bf2fd0592df999f913dd96be981cdabc0b557ca

File: ports/esp8266/boards/GENERIC_512K/manifest.py (hunk 4 above)
Comment (jimmo) [reply ports/esp8266/boards/GENERIC_512K/manifest.py:2]: @daviesian @dpgeorge This file explicitly overrides the one in the parent directory (it doesn't `include()` the parent one, like for example `GENERIC/manifest.py` does).

The idea being that the 512k "board" includes fewer modules by default, so it doesn't automatically get the contents of the modules dir.

Comment (jimmo) [reply ports/esp8266/boards/GENERIC_512K/manifest.py:2]: Added a fix in #9037

File: ports/rp2/mbedtls/mbedtls_config.h (hunk 6 above)
Comment (iabdalkader) [reply ports/rp2/mbedtls/mbedtls_config.h]: FYI @dpgeorge if cadata is in DER format it can contain only one certificate, so in other words without PEM parsing enabled (and PEM data) this function can't load a cert chain (probably should use `mbedtls_x509_crt_parse_der` instead and load one DER at a time).
```
 *                 For certificates in PEM encoding, this may be a concatenation
 *                 of multiple certificates; for DER encoding, the buffer must
 *                 comprise exactly one certificate.
 * int mbedtls_x509_crt_parse( mbedtls_x509_crt *chain, const unsigned char *buf, size_t buflen );
```
